website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,530 @@
+---
+title: "How MongoDB compares arrays in filtering"
+slug: ferretdb-compare-arrays
+authors: [chi]
+description: In this article, we explore in detail how MongoDB BSON arrays are compared
+in filter operators.
+keywords:
+  - MongoDB sorting
+  - MongoDB array comparison
+image: /img/blog/ferretdb-how-mongodb-compares-arrays.jpg
+unlisted: true
+---
+
+In this article, we explore in detail how MongoDB BSON arrays are compared in filter operators.
+
+![How MongoDB compares arrays](/img/blog/ferretdb-how-mongodb-compares-arrays.jpg)
+
+<!--truncate-->
+
+## Understanding BSON Comparison for Filtering
+
+The purpose of comparing BSON values is to determine which value is
+equal, greater or less than the other.
+The comparison is used in various operations, such as sorting, filter operators,
+as well as update operators.
+When comparing values of different BSON types, the [BSON Comparison Order](#bson-comparison-order)
+is used to determine the result.
+However, in the case of array comparison against other BSON value,
+the comparison may use the BSON comparison order or an element from the array.
+
+In this blog post, we will discuss the comparison of BSON values used in filter operators,
+with a focus on arrays.
+We will provide examples to help clarify the process.
+
+### BSON Comparison Order
+
+BSON values have a defined order of comparison, listed from lowest to highest as follows:
+
+1. Null
+2. Numbers
+3. String
+4. Object
+5. Array
+6. BinData
+7. ObjectId
+8. Boolean
+9. Date
+10. Timestamp
+11. Regular Expression
+
+The BSON comparison order is used for comparing different BSON types.
+A BSON array
+may use an element from an array to compare with other BSON type.
+We will discuss in detail in the following section.
+
+All number types are the same BSON comparison order, regardless of
+`double`, `int32`, `int64` or `decimal`.
+
+### Comparison Rules of Filter
+
+When comparing an array filter argument with array collections,
+each index is iterated in order.
+At each index, BSON comparison order is used then values are compared.
+If the values at a particular index are not equal, that result is used as the final comparison result.
+If there are no more elements in the array during the iteration,
+then that array is considered smaller.
+
+When using a filter argument comparing to an array,
+an element from that array would be selected.
+This is the case for both implicit operator and `$eq` operator.
+
+For operators `$lt` and `$lte` the smallest element from the array in the collection
+is used to compare against the filter argument.
+For operators `$gt` and `$gte` the largest element is used for comparison instead.
+Additionally, `$lt`, `$gt`, `$lte` and `$gte` uses BSON comparison order
+to only select elements of the same BSON comparison order as the filter argument.
+For example if the filter argument is string, only strings are selected.
```

Comment:
Wow, seems strange. Congrats for this finding!
I wonder how does it work with corner cases. For example what if one array does not contain any value of the filter type?

Reply:
It is not matched, as the array without same BSON value works as an empty array. There is an example already, I need to highlight that in the wording explicitly.